Basilisk master bc2 bjs2 1474

TrueSamedi opened the pull request.

TrueSamedi committed:
add RecommendationRequestDto
add RecommendationRequestController
add RecommendationRequestService (not full)

TrueSamedi committed:
add RecommendationRequestMapper
add SkillRequestDto
fixed RecommendationRequestDto
fixed RecommendationRequestService

TrueSamedi committed:
add RecommendationRequestServiceTest
add RecommendationRequestControllerTest
fixed RecommendationRequestDto
fixed RecommendationRequestService
fixed RecommendationRequestM
fixed RecommendationRequestService

TrueSamedi committed:
delete SkillRequestDto
add MessageRequestException
add RequestTimeOutException
add SkillsNotFoundException
add UserNotFoundException
fixed RecommendationRequestDto
fixed RecommendationRequestService
fixed RecommendationRequestMapper
fixed RecommendationRequestController
fixed RecommendationRequestControllerTest
fixed RecommendationRequestServiceTest

TrueSamedi committed:
add SkillRequestDto
add SkillRequestMapper
fixed RecommendationRequestDto
fixed RecommendationRequestMapper
fixed RecommendationRequestController
fixed RecommendationRequestService
fixed RecommendationRequestControllerTest
fixed RecommendationRequestServiceTest

TrueSamedi committed:
add SkillRequestDto
add SkillRequestMapper
fixed RecommendationRequestDto
fixed RecommendationRequestMapper
fixed RecommendationRequestController
fixed RecommendationRequestService
fixed RecommendationRequestControllerTest
fixed RecommendationRequestServiceTest

TrueSamedi committed:
fixed RecommendationRequestService

TrueSamedi committed:
add RequestNotFoundException
fixed RecommendationRequestService
fixed RecommendationRequestController
fixed RecommendationRequestServiceTest

TrueSamedi committed:
add FilterRecommendationRequest
add FilterRecommendationRequestCreateAt
add FilterRecommendationRequestId
add FilterRecommendationRequestMessage
add FilterRecommendationRequestReceiverId
add FilterRecommendationRequestRequesterId
add FilterRecommendationRequestSkills
add FilterRecommendationRequestStatus
add FilterRecommendationRequestUpdateAt
add RequestFilterDto
fixed RecommendationRequestController
fixed RecommendationRequestDto
fixed RecommendationRequestMapper
fixed RecommendationRequestService
fixed RecommendationRequestRepository

TrueSamedi committed:
fixed RecommendationRequestService
fixed RecommendationRequestServiceTest

TrueSamedi committed:
fixed FilterRecommendationRequest
fixed FilterRecommendationRequestCreateAt
fixed FilterRecommendationRequestId
fixed FilterRecommendationRequestMessage
fixed FilterRecommendationRequestReceiverId
fixed FilterRecommendationRequestRequesterId
fixed FilterRecommendationRequestSkills
fixed FilterRecommendationRequestStatus
fixed FilterRecommendationRequestUpdateAt
fixed RecommendationRequestService
fixed RecommendationRequestServiceTest
fixed RecommendationRequestControllerTest

TrueSamedi committed:
add RejectFailException
add RejectionDto
fixed RecommendationRequestController
fixed RecommendationRequestService
fixed RecommendationRequestServiceTest

Rik-102 commented on src/main/java/school/faang/user_service/controller/recommendation/RecommendationRequestController.java:
```diff
@@ -0,0 +1,37 @@
+
+import java.util.List;
+
+@Component
+@RequiredArgsConstructor
+public class RecommendationRequestController {
+
+    private RecommendationRequestService recommendationRequestService;
```
Бин не внедрится

TrueSamedi replied:
не совсем понял, почему?

@Autowaired вроде как необязателен.

TrueSamedi replied:
аааа. ты про final? =))
чет да, проглядел я))

Rik-102 commented on src/main/java/school/faang/user_service/controller/recommendation/RecommendationRequestController.java:
```diff
@@ -0,0 +1,37 @@
+        return recommendationRequestService.create(recommendationRequest);
+    }
+
+    public RecommendationRequestDto getRecommendationRequest(long id) {
+        return recommendationRequestService.getRequest(id);
+    }
```
Понимаю, что в сервисе у тебя вызывается метод файнд.... 
который возвращает оптионал и если что кинет ошибку. 

но я бы  в этих методах  где принимается Id , сделал бы валидацию но 0 и отриц значения.
Чтобы в сервис не ходить!

Что Паша скажет по этому  поводу, интересно)

Rik-102 commented on src/main/java/school/faang/user_service/service/RecommendationRequestService.java:
```diff
@@ -0,0 +1,91 @@
+                .toDto(findRequestById(id)
+                        .orElseThrow(() -> new RequestNotFoundException("Request not found by id: " + id)));
+    }
+
+    public List<RecommendationRequestDto> getRequest(RequestFilterDto filter) {
+        Stream<RecommendationRequest> recommendationRequests = recommendationRequestRepository.findAll().stream();
+        for (FilterRecommendationRequest filterRecommendationRequest : filterRecommendationRequestList) {
+            if (filterRecommendationRequest.isApplicable(filter))
+                recommendationRequests = filterRecommendationRequest.apply(recommendationRequests, filter);
+        }
+        List<RecommendationRequestDto> recommendationRequestDtos = new ArrayList<>();
+        for (RecommendationRequest requests : recommendationRequests.toList()) {
+            recommendationRequestDtos.add(recommendationRequestMapper.toDto(requests));
+        }
+        return recommendationRequestDtos;
+    }
```
ИМХО,
я бы тут сделал по другому. Чтобы  фильтр принимал и возвращал лист. 
Да бы не гонять туда сюда: из листа в стрим из стрима в лист, а работал бы с листом!

TrueSamedi replied:
да, я изначально так же сделал(у меня в маппере даже метод остался), но так как будто более читабельно показалось)) тоже ИМХО

Rik-102 commented on src/main/java/school/faang/user_service/service/RecommendationRequestService.java:
```diff
@@ -0,0 +1,91 @@
+        return recommendationRequestMapper
+                .toDto(findRequestById(id)
+                        .orElseThrow(() -> new RequestNotFoundException("Request not found by id: " + id)));
+    }
+
+    public List<RecommendationRequestDto> getRequest(RequestFilterDto filter) {
+        Stream<RecommendationRequest> recommendationRequests = recommendationRequestRepository.findAll().stream();
+        for (FilterRecommendationRequest filterRecommendationRequest : filterRecommendationRequestList) {
+            if (filterRecommendationRequest.isApplicable(filter))
+                recommendationRequests = filterRecommendationRequest.apply(recommendationRequests, filter);
+        }
+        List<RecommendationRequestDto> recommendationRequestDtos = new ArrayList<>();
+        for (RecommendationRequest requests : recommendationRequests.toList()) {
+            recommendationRequestDtos.add(recommendationRequestMapper.toDto(requests));
+        }
+        return recommendationRequestDtos;
```
Т.к. первый Фор отфильтровал поток. То второй Фор можно заменить на стрим. 
И получится в 3 строчки (возврат , преобразование и список). 
И не нужно будет создавать лист. )

ИМХО

InnMish replied:
Если сработает так, то можно, угу

TrueSamedi committed:
fixed RecommendationRequestController

PavelAP96 commented on src/main/java/school/faang/user_service/controller/recommendation/RecommendationRequestController.java:
```diff
@@ -0,0 +1,37 @@
+    private final RecommendationRequestService recommendationRequestService;
+
+    public RecommendationRequestDto requestRecommendation(RecommendationRequestDto recommendationRequest) {
+        if ((recommendationRequest.getMessage() == null) || recommendationRequest.getMessage().isBlank() || recommendationRequest.getMessage().isEmpty())
```
recommendationRequest.getMessage().isEmpty() избыточно
проверка isBlank делает практически тоже самое, что и isEmpty, только еще проверяет, что строка не является строкой, которая содержит только пробелы

PavelAP96 commented on src/main/java/school/faang/user_service/service/RecommendationRequestService.java:
```diff
@@ -0,0 +1,91 @@
+    private final List<FilterRecommendationRequest> filterRecommendationRequestList;
+
+    public RecommendationRequestDto create(RecommendationRequestDto recommendationRequest) {
+        if (recommendationRequest.getRequesterId() == null || recommendationRequest.getReceiverId() == null)
```
здесь ты не проверяешь, что requester и reciever присутствуют в базе данных
ты просто проверяешь, что у тебя в запросе значения не null
необходимо сделать проверку на наличие их в бд

TrueSamedi replied:
переделал. тогда получается мы валидацию на null переносим всегда в контролер и конкретно в сервисе оставляем бизнесовую логику(и валидацию если потребуется)?

PavelAP96 commented on src/main/java/school/faang/user_service/service/RecommendationRequestService.java:
```diff
@@ -0,0 +1,91 @@
+        if (recommendationRequest.getRequesterId() == null || recommendationRequest.getReceiverId() == null)
+            throw new UserNotFoundException("User not found");
+        if (recommendationRequest.getCreatedAt().plusMonths(6).isAfter(recommendationRequest.getUpdatedAt())) {
+            throw new RequestTimeOutException("Last request was less than 6 months ago");
```
на самом деле не обязательно постоянно создавать новые исключения на каждый кейс, который ты обрабатываешь
уже создано достаточно много исключений, которые можно переиспользовать
например, в данном случае тебе можно использовать IllegalArgumentException с тем же сообщением, так как у тебя действительно на вход поданы данные, с которыми ты не можешь успешно завершить процесс
плодить все новые и новые исключения не нужно
только если действительно нет уже созданного исключения, которое бы тебе подходило

PavelAP96 commented on src/main/java/school/faang/user_service/service/RecommendationRequestService.java:
```diff
@@ -0,0 +1,91 @@
+                    });
+        } catch (NullPointerException e) {
+            throw new SkillsNotFoundException("Skills not found");
+        }
```
твоя конструкция с try - catch не принесет желаемого результата
у тебя тут NullPointerException может выскочить только, если в recommendationRequest skillsId будет null, но это не значит, что у тебя скиллы не найдены в бд, а то, что у тебя запрос не валидный и скиллы не были переданы

я бы на твоем месте сделал отдельный метод, который будет у тебя валидировать запрос RecommendationRequestDto recommendationRequest и проверять, что все нужные поля у тебя не пустые

будешь вызывать его в основном методе в самом начале, а дальше прописывать оставшуюся логику, не перживая, что какое - то из полей пустое

PavelAP96 commented on src/main/java/school/faang/user_service/service/RecommendationRequestService.java:
```diff
@@ -0,0 +1,91 @@
+        recommendationRequest.getSkills()
+                .forEach(skillRequestId -> skillRequestRepository.create(skillRequestId.getId(), skillRequestId.getSkillId()));
+
+        return recommendationRequestMapper
```
чет ты все в одно налипил, выглядит не очень и тяжело читать
лучше сначала сохрани запрос в базу, сохрани в переменную
а потом вызови return с маппером, в который передашь эту переменную

PavelAP96 commented on src/main/java/school/faang/user_service/controller/recommendation/RecommendationRequestController.java:
```diff
@@ -0,0 +1,37 @@
+        return recommendationRequestService.getRequest(id);
+    }
+
+    public List<RecommendationRequestDto> getRecommendationRequest(RequestFilterDto filter) {
```
ты возвращаешь список, а в название метода в единственном числе 
лучше getRecommendationRequest**s**

PavelAP96 commented on src/main/java/school/faang/user_service/filter/FilterRecommendationRequest.java:
```diff
@@ -0,0 +1,14 @@
+
+import java.util.stream.Stream;
+
+public interface FilterRecommendationRequest {
```
так как filter написано в начале названия, оно выглядит, как название метода, который должен что - то отфильтровать, а не название класса 
я бы рекомендовал назвать `RecommendationRequestFilter`
ну и соответственно во всех реализациях этого интерфейса тоже поменять

PavelAP96 commented on src/main/java/school/faang/user_service/filter/recommendation/FilterRecommendationRequestStatus.java:
```diff
@@ -0,0 +1,21 @@
+
+    @Override
+    public Stream<RecommendationRequest> apply(Stream<RecommendationRequest> recommendationRequestStream, RequestFilterDto filterDto) {
+        return recommendationRequestStream.filter(request -> request.getStatus().equals(filterDto.getStatus()));
```
фильтр не будет работать корректно, ты сравниваешь объект типа RequestStatus с String
надо поправить сравнение

PavelAP96 commented on src/main/java/school/faang/user_service/filter/recommendation/FilterRecommendationRequestSkills.java:
```diff
@@ -0,0 +1,21 @@
+
+    @Override
+    public Stream<RecommendationRequest> apply(Stream<RecommendationRequest> recommendationRequestStream, RequestFilterDto filterDto) {
+        return recommendationRequestStream.filter(request -> request.getSkills().equals(filterDto.getSkills()));
```
тут ты сравниваешь  'List<SkillRequest>' и 'List<SkillRequestDto>, что некорректно
также лучше сравнивать списки через containsAll

TrueSamedi replied:
с этим моментом возникли сложности. 
запушу к чему пришел, но у меня есть очень смутные сомнения что правильно понял

PavelAP96 commented on src/main/java/school/faang/user_service/service/RecommendationRequestService.java:
```diff
@@ -0,0 +1,91 @@
+    }
+
+    public List<RecommendationRequestDto> getRequest(RequestFilterDto filter) {
+        Stream<RecommendationRequest> recommendationRequests = recommendationRequestRepository.findAll().stream();
+        for (FilterRecommendationRequest filterRecommendationRequest : filterRecommendationRequestList) {
+            if (filterRecommendationRequest.isApplicable(filter))
+                recommendationRequests = filterRecommendationRequest.apply(recommendationRequests, filter);
+        }
+        List<RecommendationRequestDto> recommendationRequestDtos = new ArrayList<>();
+        for (RecommendationRequest requests : recommendationRequests.toList()) {
+            recommendationRequestDtos.add(recommendationRequestMapper.toDto(requests));
+        }
```
очень красиво реализуется через StreamApi и лямбды
если не получится, то глянь в разборе Влада

TrueSamedi replied:
он изначально был как в видео у Влада, но не получилось протестить. по этому переписал так чтоб в тесте использовать capture на toDto

private ArgumentCaptor<RecommendationRequest> captor;
verify(recommendationRequestMapper, times(1)).toDto(captor.capture());

но в целом согласен, могу переделать обратно))

TrueSamedi committed:
delete UserNotFoundException
fixed SkillRequestMapper
fixed RecommendationRequestServiceTest
fixed RecommendationRequestService
fixed RecommendationRequestFilterUpdateAt
fixed RecommendationRequestFilterStatus
fixed RecommendationRequestFilterSkills
fixed RecommendationRequestFilterRequesterId
fixed RecommendationRequestFilterReceiverId
fixed RecommendationRequestFilterMessage
fixed RecommendationRequestFilterId
fixed RecommendationRequestFilterCreateAt
fixed RecommendationRequestFilter
fixed RecommendationRequestController

TrueSamedi committed:
fixed RecommendationRequestService
fixed RecommendationRequestController
fixed RecommendationRequestDto
fixed RejectionDto

TrueSamedi committed:
fixed SkillRequestMapper

TrueSamedi committed:
fixed RecommendationRequestService
fixed RecommendationRequestRepository

InnMish commented on src/main/java/school/faang/user_service/controller/recommendation/RecommendationRequestController.java:
```diff
@@ -0,0 +1,50 @@
+@Slf4j
+@RestController
+@RequiredArgsConstructor
+@RequestMapping("/api/v1/recommendationRequest")
```
/api/v1 вынеси в конфигурационный файл 
Это не часть маппинга конкретно этого контроллера, а общее значение в целом для всего приложения 
Типа, микросервис первой версии 

И чтобы в каждом контроллере не писать одно и то же, можно применить сразу для всех в конфиге 

настройка там типа 
spring
mvc
servlet
path - что-то такое

TrueSamedi replied:
мы с командой вроде как договорились не в конфиге это писать)

InnMish commented on src/main/java/school/faang/user_service/controller/recommendation/RecommendationRequestController.java:
```diff
@@ -0,0 +1,50 @@
+
+    private final RecommendationRequestService recommendationRequestService;
+
+    @PostMapping("/create")
```
Такой маппинг - не по ресту 
У тебя сам метод Post уже говорит о том, что ты что-то отправляешь в сервис, чтобы с сущностью поработать 
достаточно будет того, что в общем маппинге уже есть 

Глагол в маппинге обычно мы не пишем, если только без него ну вообще не понять, что происходит

InnMish commented on src/main/java/school/faang/user_service/controller/recommendation/RecommendationRequestController.java:
```diff
@@ -0,0 +1,50 @@
+
+    @PostMapping("/create")
+    public RecommendationRequestDto requestRecommendation(@RequestBody RecommendationRequestDto recommendationRequest) {
+        if ((recommendationRequest.getMessage() == null) || recommendationRequest.getMessage().isBlank())
```
а двойные скобки тут точно нужны?

InnMish commented on src/main/java/school/faang/user_service/controller/recommendation/RecommendationRequestController.java:
```diff
@@ -0,0 +1,50 @@
+    @PostMapping("/create")
+    public RecommendationRequestDto requestRecommendation(@RequestBody RecommendationRequestDto recommendationRequest) {
+        if ((recommendationRequest.getMessage() == null) || recommendationRequest.getMessage().isBlank())
+            throw new MessageRequestException("Incorrect user's message");
```
У тебя что, под каждый ошибки свой класс?) 
Это не очень хорошая практика, какой-то микроконтроль получается 
Я бы создала одно общее исключение, типа DataValidationException, и все ошибки валидации в него складывала

InnMish commented on src/main/java/school/faang/user_service/controller/recommendation/RecommendationRequestController.java:
```diff
@@ -0,0 +1,50 @@
+    }
+
+    @GetMapping("/{recommendationRequestId}")
+    public RecommendationRequestDto getRecommendationRequest(@PathVariable("recommendationRequestId") long id) {
```
если назовешь само поле так, как у тебя в маппинг передается, можно будет в кавычках к PathVariable его не писать - но эт мелочь

InnMish commented on src/main/java/school/faang/user_service/controller/recommendation/RecommendationRequestController.java:
```diff
@@ -0,0 +1,50 @@
+    public RecommendationRequestDto rejectRequest(@PathVariable("requestId") long id, @RequestBody RejectionDto rejection) {
+        return recommendationRequestService.rejectRequest(id, rejection);
+    }
+
```
строчку убираем-с

InnMish commented on src/main/java/school/faang/user_service/dto/RecommendationRequestDto.java:
```diff
@@ -0,0 +1,25 @@
+    private List<SkillRequestDto> skills;
+    private Long requesterId;
+    private Long receiverId;
+    private LocalDateTime createdAt = LocalDateTime.now();
```
а зачем тут проставляешь? 
Хибер сам умеет проставлять время, если в сущности над этим полем висит аннотация CreationTimestamp - проверь в бд без вот этого проставления ручного - изменится или нет

InnMish commented on src/main/java/school/faang/user_service/dto/RecommendationRequestDto.java:
```diff
@@ -0,0 +1,25 @@
+    private Long id;
+    private String message;
+    private RequestStatus status = RequestStatus.PENDING;
+    private List<SkillRequestDto> skills;
```
В дто лучше всего работать с айдишниками только 

По-первых, потому что заполнять всю дто долго и тянуть много данных приходится, во-вторых, потому что любую сущность ты при необходимости из бд можешь получить, тебе не надо от пользователя ВООБЩЕ ВСЕ данные получать. Берешь только необходимые 
Поэтому тут сделать список лонгов - айдишников

InnMish commented on src/main/java/school/faang/user_service/dto/RequestFilterDto.java:
```diff
@@ -0,0 +1,20 @@
+    private Long id;
+    private String message;
+    private String status;
+    private List<SkillRequestDto> skills;
```
список айдишников

InnMish commented on src/main/java/school/faang/user_service/dto/RequestFilterDto.java:
```diff
@@ -0,0 +1,20 @@
+@Data
+public class RequestFilterDto {
+
+    private Long id;
```
точно хочешь фильтровать по айди? Это ж уникально значение)

InnMish commented on src/main/java/school/faang/user_service/dto/RequestFilterDto.java:
```diff
@@ -0,0 +1,20 @@
+public class RequestFilterDto {
+
+    private Long id;
+    private String message;
```
по месседжу я бы тоже не фильтровала - геморно, только если у нас не какой-то супер-крутой поиск по словам

InnMish commented on src/main/java/school/faang/user_service/filter/recommendation/RecommendationRequestFilterCreateAt.java:
```diff
@@ -0,0 +1,21 @@
+
+    @Override
+    public Stream<RecommendationRequest> apply(Stream<RecommendationRequest> recommendationRequestStream, RequestFilterDto filterDto) {
+        return recommendationRequestStream.filter(request -> request.getCreatedAt().equals(filterDto.getCreatedAt()));
```
хехехе 
А вот у тебя тип этих полей LocalDateTime 
Ты смотрел, как они выглядят? Там время с миллисекундами вроде 

У тебя условие на equals никогда не сработает. Сделай проверку на isBefore или isAfter

InnMish commented on src/main/java/school/faang/user_service/filter/recommendation/RecommendationRequestFilterId.java:
```diff
@@ -0,0 +1,21 @@
+    }
+
+    @Override
+    public Stream<RecommendationRequest> apply(Stream<RecommendationRequest> recommendationRequestStream, RequestFilterDto filterDto) {
+        return recommendationRequestStream.filter(request -> request.getId() == filterDto.getId());
```
вот это выпилить, не будем так фильтровать

InnMish commented on src/main/java/school/faang/user_service/filter/recommendation/RecommendationRequestFilterMessage.java:
```diff
@@ -0,0 +1,21 @@
+    }
+
+    @Override
+    public Stream<RecommendationRequest> apply(Stream<RecommendationRequest> recommendationRequestStream, RequestFilterDto filterDto) {
+        return recommendationRequestStream.filter(request -> request.getMessage().equals(filterDto.getMessage()));
```
и так тоже не будем

InnMish commented on src/main/java/school/faang/user_service/filter/recommendation/RecommendationRequestFilterUpdateAt.java:
```diff
@@ -0,0 +1,21 @@
+    }
+
+    @Override
+    public Stream<RecommendationRequest> apply(Stream<RecommendationRequest> recommendationRequestStream, RequestFilterDto filterDto) {
+        return recommendationRequestStream.filter(request -> request.getUpdatedAt().equals(filterDto.getUpdatedAt()));
```
как и в первом комменте про время

InnMish commented on src/main/java/school/faang/user_service/service/RecommendationRequestService.java:
```diff
@@ -0,0 +1,107 @@
+    private final SkillRequestRepository skillRequestRepository;
+    private final UserRepository userRepository;
+    private final RecommendationRequestMapper recommendationRequestMapper;
+    private final SkillRequestMapper skillRequestMapper;
```
не используется

InnMish commented on src/main/java/school/faang/user_service/service/RecommendationRequestService.java:
```diff
@@ -0,0 +1,107 @@
+
+        recommendationRequest.getSkills()
+                .forEach(skillRequestId -> {
+                    if (!skillRequestRepository.existsById(skillRequestId.getSkillId()))
```
а что это тут? не поняла логику 
Тебе здесь надо проверить, что в бд в таблице skill есть такие навыки вообще, которые пользователь пытается запросить в этом запросе рекомендации 

Для этого тебе не нужно лезть в таблицу skillRequest, надо только проверить именно в skillRepository, что все скиллы существуют вообще в природе)

InnMish commented on src/main/java/school/faang/user_service/service/RecommendationRequestService.java:
```diff
@@ -0,0 +1,107 @@
+        recommendationRequest.getSkills()
+                .forEach(skillRequestId -> {
+                    if (!skillRequestRepository.existsById(skillRequestId.getSkillId()))
+                        skillsCheck.getAndIncrement();
```
и вот это тоже не поняла.. почему атомик? Мы ж просто проверяем количество

InnMish commented on src/main/java/school/faang/user_service/service/RecommendationRequestService.java:
```diff
@@ -0,0 +1,107 @@
+                        skillsCheck.getAndIncrement();
+                });
+
+        if (skillsCheck.longValue() > 0)
+            throw new SkillsNotFoundException("Skills not found");
```
шиворот-навыворот получилось 
Если насчитали много, значит ничего нет 

Правильнее же будет "если ноль, значит ничего нет"))

InnMish replied:
Ну и исключение вот это замени на единое DataValidationException

InnMish commented on src/main/java/school/faang/user_service/service/RecommendationRequestService.java:
```diff
@@ -0,0 +1,107 @@
+        if (skillsCheck.longValue() > 0)
+            throw new SkillsNotFoundException("Skills not found");
+
+        recommendationRequest.getSkills()
+                .forEach(skillRequestId -> skillRequestRepository.create(skillRequestId.getId(), skillRequestId.getSkillId()));
```
skillRequestRepository.create должен вызываться в сервисе SkillRequestService или просто в SkillService 
Этот метод может еще в каком-то месте понадобиться, поэтому было бы логичнее его вынести в отдельный ему соответствующий сервис 

Плюс сингл респонсибилити - твой сервис рекомендаций работает только с ними, а сервис скиллов уже работает со скиллами

InnMish commented on src/main/java/school/faang/user_service/service/RecommendationRequestService.java:
```diff
@@ -0,0 +1,107 @@
+    public RecommendationRequestDto getRequest(long id) {
+        return recommendationRequestMapper
+                .toDto(findRequestById(id)
+                        .orElseThrow(() -> new RequestNotFoundException("Request not found by id: " + id)));
```
Тут экспшен лучше назвать EntityNotFoundException - чтобы потом переиспользовать 
А то такое название слишком заточено именно под твою сущность

InnMish commented on src/main/java/school/faang/user_service/service/RecommendationRequestService.java:
```diff
@@ -0,0 +1,107 @@
+    }
+
+    public RecommendationRequestDto rejectRequest(long id, RejectionDto rejection) {
+        RecommendationRequest request = findRequestById(id).orElseThrow(() -> new RequestNotFoundException("Request not found by id: " + id));
```
Но-но 
Ошибку вот эту кидай сразу в методе findRequestById 
Она же постоянная у тебя, если сущность не была найдена 
Зачем из метода того возвращать optional, если можно обработать его сразу и вернуть сущность

InnMish commented on src/main/java/school/faang/user_service/service/RecommendationRequestService.java:
```diff
@@ -0,0 +1,107 @@
+            request.setRejectionReason(rejection.getReason());
+            request.setStatus(RequestStatus.REJECTED);
+            recommendationRequestRepository.save(request);
+        } else {
+            throw new RejectFailException("Request is accepted or rejected");
```
validationexception

InnMish commented on src/main/java/school/faang/user_service/service/RecommendationRequestService.java:
```diff
@@ -0,0 +1,107 @@
+    }
+
+    private void validate(RecommendationRequestDto recommendationRequest) {
+        if (!userRepository.existsById((recommendationRequest.getRequesterId())) || !userRepository.existsById(recommendationRequest.getReceiverId()))
```
обращение к юзер репозиторию перенести в юзер сервис и сделать это единым методом 
Который ты тут потом два раза вызовешь с разными параметрами здесь

InnMish commented on src/main/java/school/faang/user_service/service/RecommendationRequestService.java:
```diff
@@ -0,0 +1,107 @@
+
+    private void validate(RecommendationRequestDto recommendationRequest) {
+        if (!userRepository.existsById((recommendationRequest.getRequesterId())) || !userRepository.existsById(recommendationRequest.getReceiverId()))
+            throw new IllegalArgumentException("User not found");
```
ну, если не найден, значит entityNotFound

InnMish commented on src/main/java/school/faang/user_service/service/RecommendationRequestService.java:
```diff
@@ -0,0 +1,107 @@
+    private void validate(RecommendationRequestDto recommendationRequest) {
+        if (!userRepository.existsById((recommendationRequest.getRequesterId())) || !userRepository.existsById(recommendationRequest.getReceiverId()))
+            throw new IllegalArgumentException("User not found");
+        if (recommendationRequest.getSkills() == null)
+            throw new IllegalArgumentException("skills is null");
```
datavalidation

InnMish commented on src/main/java/school/faang/user_service/service/RecommendationRequestService.java:
```diff
@@ -0,0 +1,107 @@
+    }
+
+}
+
```
лишние строчки убрать

InnMish commented on src/test/java/school/faang/user_service/service/controller/recommendation/RecommendationRequestControllerTest.java:
```diff
@@ -0,0 +1,41 @@
+
+    @Test
+    void testMessageNotNull() {
+        Assert.assertThrows(MessageRequestException.class, ()-> recommendationRequestController.requestRecommendation(new RecommendationRequestDto(5L, null, "status", new ArrayList<SkillRequestDto>(), 5L, 6L, LocalDateTime.now(), LocalDateTime.now().plusMonths(7))));
```
а что такие гусеницы длинные? Перенеси на след строчку 
В идее же есть вертикальная полосочка-разделитель 
Вот на нее ориентируйся, чтобы код не выезжал

InnMish commented on src/test/java/school/faang/user_service/service/service/RecommendationRequestServiceTest.java:
```diff
@@ -0,0 +1,168 @@
+        recommendationRequestService.rejectRequest(7L, rejectionDto);
+        verify(recommendationRequestMapper, times(1)).toDto(captor.capture());
+        assertEquals(RequestStatus.REJECTED, captor.getValue().getStatus());
+    }
```
еще бы тестов на фильтры написать 
и на ошибки разные

TrueSamedi committed:
fixed RecommendationRequestController

TrueSamedi committed:
add DataValidationException
fixed RecommendationRequestController
fixed RecommendationRequestControllerTest
fixed RecommendationRequestService
fixed RecommendationRequestServiceTest
delete MessageRequestException
delete RequestNotFoundException
delete RequestTimeOutException.java
delete SkillsNotFoundException.java

TrueSamedi committed:
fixed RecommendationRequestDto
fixed RecommendationRequestServiceTest
delete RecommendationRequestFilterId.java
delete RecommendationRequestFilterMessage.java

TrueSamedi committed:
add UserService
fixed RecommendationRequestController
fixed RecommendationRequestFilterCreateAt
fixed RecommendationRequestFilterUpdateAt
fixed RecommendationRequestService
fixed RecommendationRequestServiceTest

Commit:
fix
